### ISSUE TITLE
`__init_subclass__` takes `cls` not `self`.

### ISSUE THREAD

**Issue by MicaelJarniac** [sourcery-ai/sourcery#220]
### Issue description or question

```py
class Foo:
    def __init_subclass__(cls) -> None:  # Sourcery - The first argument to instance methods should be `self`.
        reveal_type(cls)  # Revealed type is "Type[Foo]" - mypy(note)
```

### Sourcery Version

`v0.11.2`

### Code editor or IDE name and version

VS Code
Version: 1.66.0
Commit: e18005f0f1b33c29e81d732535d8c0e47cafb0b5
Date: 2022-03-30T05:50:41.156Z
Electron: 17.2.0
Chromium: 98.0.4758.109
Node.js: 16.13.0
V8: 9.8.177.11-electron.0
OS: Linux x64 5.15.28-1-MANJARO snap

### OS name and version

Operating System: Manjaro Linux
KDE Plasma Version: 5.24.3
KDE Frameworks Version: 5.91.0
Qt Version: 5.15.3
Kernel Version: 5.15.28-1-MANJARO (64-bit)
Graphics Platform: X11
Processors: 4 × AMD Ryzen 3 2200G with Radeon Vega Graphics
Memory: 15.6 GiB of RAM
Graphics Processor: NVIDIA GeForce GTX 970/PCIe/SSE2

**Comment by diceroll123** [comment]
Same for `__new__`, possibly more false positives.

**Comment by bm424** [comment]
Hi @MicaelJarniac, thanks for raising this. Just had a little check alongside `__init_subclass__` as you point out, and `__new__` (thanks @diceroll123), we also need to use `cls` instead of `self` for `__class_getitem__`. Any other missing methods warmly welcomed.

**Comment by MicaelJarniac** [comment]
I wonder if it'd be possible to run some kind of type-checker while making those suggestions.

In my example, Mypy (with `reveal_type`) does correctly know that the type of the first argument is `Type[Foo]`, and not `Foo`.

If Sourcery were to run a type-checker, be it Mypy or any other, then the job of knowing whether the first argument should be `self` or `cls` could be off-sourced to it.

I believe most, if not all, type-checkers will already know exactly what the first argument is.

**Comment by MicaelJarniac** [comment]
And if not running a type-checker on Sourcery runtime, I believe a type-checker could be used for training Sourcery beforehand.

Worst case, I believe it'd be possible to generate a Python file with all "dunder" methods and `reveal_type` inside of them, run a type-checker on that file, parse its output, and use that for configuring Sourcery.

But I believe it'd be possible to do something smarter than that. There's likely a list of known "dunders" and what their arguments are in the source of those type-checkers, so that list could be parsed directly, without even having to run the type-checkers.

**Comment by MicaelJarniac** [comment]
https://github.com/python/mypy/blob/ab1b48808897d73d6f269c51fc29e5c8078fd303/mypy/semanal.py#L702

https://github.com/python/mypy/blob/ab1b48808897d73d6f269c51fc29e5c8078fd303/mypy/stubtest.py#L328

https://github.com/python/mypy/blob/ab1b48808897d73d6f269c51fc29e5c8078fd303/mypy/stubtest.py#L439

**Comment by Hellebore** [comment]
Fixed in 0.11.3
